### DISCUSSION TITLE
Fields in JSON with the same name

### DISCUSSION THREAD

**Discussion by pjedynak** [nlohmann/json#3567] [Q&A] [answered]
Hi,
I'm wondering what is correct behavior if input json file contains several fields with the same. It is invalid json file, so I would assume that during parsing I should get some error. But unfortunately I see that only last is visible and previous ones are overridden.
Below some part of the code:
```
#include <nlohmann/json.hpp>
#include <sstream>
#include <iostream>

int main()
{
    std::stringstream ss {};
    ss << R"({"item1":1, "item2": 2, "item1":4})";
    nlohmann::json j{};
    ss >> j;
    std::cout << j.dump() << std::endl;
};
```
Result of dump is `{"item1":4,"item2":2}`. Maybe I should use another approach for parsing json format with verification of duplicates ?

**Comment by falbrechtskirchinger** [comment, marked as answer]
Duplicate keys are perfectly valid but not recommended (see footnotes below) since the behavior is not standardized and varies between implementations.
You can use [`basic_json::sax_parse()`](https://json.nlohmann.me/api/basic_json/sax_parse/) to check for duplicate keys either during parsing or as a verification step before calling `basic_json::parse()` (the latter requiring a lot less work).

You can find an example here: https://godbolt.org/z/f1h9PEaGM

```cpp
#include <nlohmann/json.hpp>
#include <iostream>
#include <vector>

using json = nlohmann::json;

class sax_validator : public json::json_sax_t {
  public:
    bool null() override { return true; }
    bool boolean(bool val) override { return true; }
    bool number_integer(number_integer_t val) override { return true; }
    bool number_unsigned(number_unsigned_t val) override { return true; }
    bool number_float(number_float_t val, const string_t& s) override { return true; }
    bool string(string_t& val) override { return true; }
    bool start_array(std::size_t elements) override { return true; }
    bool end_array() override { return true; }
    bool binary(json::binary_t& val) override { return true; }

    bool parse_error(std::size_t position, const std::string& last_token, const json::exception& ex) override {
        throw ex;
        return false;
    }

    bool start_object(std::size_t elements) override {
        m_stack.emplace_back();
        return true;
    }

    bool end_object() override {
        m_stack.pop_back();
        return true;
    }

    bool key(string_t& val) override
    {
        auto &obj = m_stack.back();
        auto it = std::lower_bound(obj.begin(), obj.end(), val);
        if(it != obj.end() && *it == val) {
            std::cout << "duplicate key: " << val << '\n';
            return false;
        }
        obj.insert(it, std::move(val));
        return true;
    }

private:
    std::vector<std::vector<std::string>> m_stack;
};

int main() {
    sax_validator sax;
    bool result = json::sax_parse(R"(
    {
        "item1": 1,
        "item2": 2,
        "item1": 4
    })", &sax);
    std::cout << std::boolalpha << "result: " << result << '\n';
}
```

<hr />
Footnotes:

https://datatracker.ietf.org/doc/html/rfc8259#section-4
> The names within an object SHOULD be unique.

https://datatracker.ietf.org/doc/html/rfc2119#section-3
> SHOULD   This word, or the adjective "RECOMMENDED", mean that there
   may exist valid reasons in particular circumstances to ignore a
   particular item, but the full implications must be understood and
   carefully weighed before choosing a different course.

**Reply by pjedynak** [reply]
Thank you very much for clarification and example.
